## NumPy İşlemleri

## Aritmetik

Dizilerle, diziler veya dizilerle, skalerler arasında aritmetik  işlemleri kolayca yapabilirsiniz.

In [1]:
import numpy as np

In [2]:
arr = np.arange(0,10)
arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [3]:
arr + 5 # dizinin elemanlarının herbirine 5 skalerini ekler

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

In [4]:
list = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] # dizinin elemanlarını çoğaltır
list * 2

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [5]:
arr + arr

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [6]:
arr * arr 

array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])

In [7]:
arr - arr

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [8]:
# Bu işlem, sıfıra bölme konusunda bir uyarı verir, 
# ancak bir hata oluşturmaz! Sadece konumu **nan** ile doldurur
arr / arr

/var/folders/_s/bc6zc6vj4ns_qtdjxhylpyww0000gn/T/ipykernel_2337/582049721.py:3: RuntimeWarning: invalid value encountered in true_divide
  arr / arr


array([nan,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])

In [9]:
# Ayrıca sonsuzla ilgili bir uyarı (ama bir hata değil) verir
1 / arr

/var/folders/_s/bc6zc6vj4ns_qtdjxhylpyww0000gn/T/ipykernel_2337/2696729746.py:2: RuntimeWarning: divide by zero encountered in true_divide
  1 / arr


array([       inf, 1.        , 0.5       , 0.33333333, 0.25      ,
       0.2       , 0.16666667, 0.14285714, 0.125     , 0.11111111])

In [10]:
arr**3

array([  0,   1,   8,  27,  64, 125, 216, 343, 512, 729])

## Evrensel Dizi Fonksiyonları

NumPy matematiksel işlem yapan birçok evrensel dizi fonksiyonuna sahiptir: (http://docs.scipy.org/doc/numpy/reference/ufuncs.html).

Yaygın olanları gösterelim:

In [12]:
# Karekök Alma
np.sqrt(arr)

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ,
       2.23606798, 2.44948974, 2.64575131, 2.82842712, 3.        ])

In [13]:
# üstel hesaplama (e^arr)
np.exp(arr)

array([1.00000000e+00, 2.71828183e+00, 7.38905610e+00, 2.00855369e+01,
       5.45981500e+01, 1.48413159e+02, 4.03428793e+02, 1.09663316e+03,
       2.98095799e+03, 8.10308393e+03])

In [14]:
# Sinüs gibi trigonometrik fonksiyonlar
np.sin(arr)

array([ 0.        ,  0.84147098,  0.90929743,  0.14112001, -0.7568025 ,
       -0.95892427, -0.2794155 ,  0.6569866 ,  0.98935825,  0.41211849])

In [15]:
# Doğal Logaritma alma
np.log(arr)

/var/folders/_s/bc6zc6vj4ns_qtdjxhylpyww0000gn/T/ipykernel_2337/443999168.py:2: RuntimeWarning: divide by zero encountered in log
  np.log(arr)


array([      -inf, 0.        , 0.69314718, 1.09861229, 1.38629436,
       1.60943791, 1.79175947, 1.94591015, 2.07944154, 2.19722458])

## Dizilerle İlgili Özet İstatistikler

NumPy, **sum, mean ve max** gibi genel özet istatistikleri de sunar. Bunlar bir dizideki yöntemler olarak çağırılır.

In [26]:
arr = np.arange(0,10)
arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [20]:
arr.sum()

45

In [21]:
arr.mean() 

4.5

In [22]:
arr.max()

9

In [27]:
arr.std()

2.8722813232690143

In [28]:
arr.var()

8.25

<strong>Diğer özet istatistikler şunları içerir::</strong>
<pre>
arr.min() returns 0                   minimum
arr.var() returns 8.25                varyans
arr.std() returns 2.8722813232690143  standard sapma
</pre>

## Eksen Mantığı
2-boyutlu dizilerle (matrisler) çalışırken satırları ve sütunları dikkate almalıyız. Dizi terimlerinde, axis=0 (satırlar boyunca) dikey eksen ve axis=1 (sütunlar boyunca) yatay eksendir. Bu değerler, <tt>arr.shape</tt> değerlerinin döndürüldüğü sıraya karşılık gelir.

Bunun, yukarıdaki özet istatistik hesaplamalarımızı nasıl etkilediğine bakalım.

In [29]:
arr_2d = np.array([[1,2,3,4],[5,6,7,8],[9,10,11,12]])
arr_2d

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [30]:
arr_2d.sum(axis=0)

array([15, 18, 21, 24])

<tt>axis=0</tt> ile dikey eksen boyuncaki matrisin elemanlarının toplamı döndürülür. Yani, <tt>[(1+5+9), (2+6+10), (3+7+11), (4+8+12)]</tt>

<img src='image/axis_logic.png' width=400/>

In [34]:
arr_2d.shape

(3, 4)

<tt>arr_2d</tt> dizisinin 3 satır, 4 sütuna sahip olduğunu söyler.

<tt>arr_2d.sum(axis=0)</tt> 'de, her satırdaki ilk eleman toplanır, ardından ikinci eleman ve bu böyle devam eder.

In [35]:
# HÜCREYİ ÇALIŞTIRMADAN ÖNCE NEYİ DÖNDÜRECEĞİNİ DÜŞÜNÜN!
arr_2d.sum(axis=1)

array([10, 26, 42])

# SON!